pkg/controller/container/nodepool.go

```diff
@@ -0,0 +1,243 @@
+func (c *NodePoolController) SetupWithManager(mgr ctrl.Manager) error {
+	r := resource.NewManagedReconciler(mgr,
+		resource.ManagedKind(v1alpha1.NodePoolGroupVersionKind),
+		resource.WithExternalConnecter(&nodePoolConnector{kube: mgr.GetClient(), newServiceFn: container.NewService}))
+
+	name := strings.ToLower(fmt.Sprintf("%s.%s", v1alpha1.NodePoolKindAPIVersion, v1alpha1.Group))
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha1.NodePool{}).
+		Complete(r)
+}
+
+type nodePoolConnector struct {
+	kube         client.Client
+	newServiceFn func(ctx context.Context, opts ...option.ClientOption) (*container.Service, error)
+}
+
+func (c *nodePoolConnector) Connect(ctx context.Context, mg resource.Managed) (resource.ExternalClient, error) {
+	i, ok := mg.(*v1alpha1.NodePool)
+	if !ok {
+		return nil, errors.New(errNotNodePool)
+	}
+
+	p := &gcpv1alpha3.Provider{}
+	if err := c.kube.Get(ctx, meta.NamespacedNameOf(i.Spec.ProviderReference), p); err != nil {
+		return nil, errors.Wrap(err, errGetProvider)
+	}
+
+	s := &corev1.Secret{}
+	n := types.NamespacedName{Namespace: p.Spec.Secret.Namespace, Name: p.Spec.Secret.Name}
+	if err := c.kube.Get(ctx, n, s); err != nil {
+		return nil, errors.Wrap(err, errGetProviderSecret)
+	}
+
+	client, err := c.newServiceFn(ctx,
+		option.WithCredentialsJSON(s.Data[p.Spec.Secret.Key]),
+		option.WithScopes(container.CloudPlatformScope))
+	return &nodePoolExternal{container: client, projectID: p.Spec.ProjectID, kube: c.kube}, errors.Wrap(err, errNewClient)
+}
+
+type nodePoolExternal struct {
+	kube      client.Client
+	container *container.Service
+	projectID string
+}
+
+func (e *nodePoolExternal) Observe(ctx context.Context, mg resource.Managed) (resource.ExternalObservation, error) { // nolint:gocyclo
+	cr, ok := mg.(*v1alpha1.NodePool)
+	if !ok {
+		return resource.ExternalObservation{}, errors.New(errNotNodePool)
+	}
+
+	existing, err := e.container.Projects.Locations.Clusters.NodePools.Get(np.GetFullyQualifiedName(cr.Spec.ForProvider, meta.GetExternalName(cr))).Context(ctx).Do()
+	if err != nil {
+		return resource.ExternalObservation{}, errors.Wrap(resource.Ignore(gcp.IsErrorNotFound, err), errGetNodePool)
+	}
+
+	cr.Status.AtProvider = np.GenerateObservation(*existing)
+	currentSpec := cr.Spec.ForProvider.DeepCopy()
+	np.LateInitializeSpec(&cr.Spec.ForProvider, *existing)
+	if !cmp.Equal(currentSpec, &cr.Spec.ForProvider) {
+		if err := e.kube.Update(ctx, cr); err != nil {
+			return resource.ExternalObservation{}, errors.Wrap(err, errManagedNodePoolUpdateFailed)
+		}
+	}
+
+	switch cr.Status.AtProvider.Status {
+	case v1alpha1.NodePoolStateRunning:
+		cr.Status.SetConditions(runtimev1alpha1.Available())
+		resource.SetBindable(cr)
+	case v1alpha1.NodePoolStateProvisioning:
+		cr.Status.SetConditions(runtimev1alpha1.Creating())
+	case v1alpha1.NodePoolStateUnspecified, v1alpha1.NodePoolStateRunningError, v1alpha1.NodePoolStateError, v1alpha1.NodePoolStateReconciling:
```

Comment:
Yeah I ran into this same issue with `v1beta1` `GKECluster`. However, I think we can mark it as `Available` because the first reconciliation should be in `Provisioning` not `Reconciling` state. If we do that here I will also update `GKECluster`.

Reply:
I feel like we should just skip this state, i.e. our condition would be left in the state right before `RECONCILING`, which should be OK but that's not a strong opinion. I could be convinced either way.